File: OJ practices/LeetCode题目集/697. Degree of an Array(easy).cpp
/*
 * @Autor: SourDumplings
 * @Date: 2019-10-21 16:34:27
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/degree-of-an-array/
 */

class Solution
{
public:
    int findShortestSubArray(vector<int> &nums)
    {
        unordered_map<int, int> freq;
        for (auto &&i : nums)
        {
            ++freq[i];
        }
        int maxF = 0;
        unordered_set<int> maxFSet;
        for (auto &&p : freq)
        {
            if (p.second > maxF)
            {
                maxF = p.second;
                maxFSet.clear();
                maxFSet.insert(p.first);
            }
            else if (p.second == maxF)
            {
                maxFSet.insert(p.first);
            }
        }
        unordered_map<int, int> firstOccur;
        unordered_map<int, int> lastOccur;
        int l = nums.size();
        int res = l;
        for (int i = 0; i < l; i++)
        {
            int num = nums[i];
            if (maxFSet.find(num) != maxFSet.end())
            {
                if (firstOccur.find(num) == firstOccur.end())
                {
                    firstOccur[num] = i;
                }
                lastOccur[num] = i;
            }
        }
        for (auto &&p : firstOccur)
        {
            int num = p.first;
            if (lastOccur[num] - p.second + 1 < res)
            {
                res = lastOccur[num] - p.second + 1;
            }
        }

        return res;
    }
};

File: OJ practices/LeetCode题目集/697. Degree of an Array(easy).java
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/*
 * @Autor: SourDumplings
 * @Date: 2019-10-21 16:58:10
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/degree-of-an-array/
 */

class Solution
{
    public int findShortestSubArray(int[] nums)
    {
        Map<Integer, Integer> count = new HashMap<>();
        Map<Integer, Integer> left = new HashMap<>();
        Map<Integer, Integer> right = new HashMap<>();
        int l = nums.length;
        for (int i = 0; i < l; i++)
        {
            int num = nums[i];
            left.putIfAbsent(num, i);
            count.merge(num, 1, Integer::sum);
            right.put(num, i);
        }
        int maxF = Collections.max(count.values());
        int res = l;
        for (int num : left.keySet())
        {
            if (count.get(num) == maxF)
            {
                int r = right.get(num);
                int le = left.get(num);
                if (r - le + 1 < res)
                {
                    res = r - le + 1;
                }
            }
        }
        return res;
    }
}